### PR TITLE
token healing impl

### LINKED ISSUE #28346: Token healing (under 40 LOC)
### Feature request

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality, though I'm not sure where it would fit as a transformers feature.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.

### Your contribution

My implementation (under 40 LOC): https://github.com/Ayenem/TokenHealer/tree/main

### PR DESCRIPTION
**Pull request by ahmed-moubtahij**
# What does this PR do?

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.

Fixes [#28346](https://github.com/huggingface/transformers/issues/28346)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

- @gante

### CODE REVIEW

**Comment by gante** [comment]
CI is failing due to an automatic update in the pytest package, we are tracking it. Will let you know when it is sorted -- it will need a rebase

**Comment by Ayenem** [comment]
image

**Comment by Ayenem** [comment]
image

**Comment by gante** [comment]
Now rebased after #29320 was merged, which was causing the last set of errors seen here. If everything went well, we should see a green CI here 🤞

**Comment by Ayenem** [comment]
> @Ayenem FYI, I've reverted the tokenizer input to your original suggestion (tokenizer passed to `generate`), after a discussion I had with @Rocketknight1. That way, the input is standardized and matches another incoming PR (#28932) 🤗

It does feel better to offload the tokenizer choice and loading to the caller. Thanks again for following up on this 🙏

**Comment by Ayenem** [comment]
CI is green! It was possible :')

**Comment by gante** [comment]
ping @ArthurZucker :)

**Comment by ArthurZucker** [comment]
Sorry for the late review on it!

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 
 import torch
 import torch.distributed as dist
+from pygtrie import CharTrie
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:25]: if this is an optinal dependency we need to protect the import

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1803,6 +1809,67 @@ def typeerror():
...
+        Return:
+            `torch.LongTensor` where each sequence has its tail token replaced with its appropriate extension.
+        """
+        if tokenizer is None:
+            raise ValueError(
+                " When generating with token healing, you must pass the model's tokenizer to the `tokenizer` "
+                "argument of `generate`."
+            )
+        bos_id, pad_id = tokenizer.bos_token_id, tokenizer.pad_token_id
+        vocab_trie = CharTrie(tokenizer.get_vocab())
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1829]: BTW we have https://github.com/huggingface/transformers/blob/main/src/transformers/tokenization_utils.py#L52 

which could be used for this? Would remove the dependency? (It's might be additional work as well)

**Review by ArthurZucker — COMMENTED** [review]
Left a few nits, mostly safely import and protect the function as the new dependency is optional / should be optional. Potentially use our own trie?

**Pull request closed without merging**